network/v1alpha1/types_dnsnameresolver.go

```diff
@@ -0,0 +1,129 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +openshift:compatibility-gen:level=4
+
+// DNSNameResolver stores the DNS name resolution information of a DNS name. It can be enabled by the TechPreviewNoUpgrade feature set.
+// It can also be enabled by the feature gate DNSNameResolver when using CustomNoUpgrade feature set.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+type DNSNameResolver struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the DNSNameResolver.
+	// +kubebuilder:validation:Required
+	Spec DNSNameResolverSpec `json:"spec"`
+	// status is the most recently observed status of the DNSNameResolver.
+	// +optional
+	Status DNSNameResolverStatus `json:"status,omitempty"`
+}
+
+// DNSNameResolverSpec is a desired state description of DNSNameResolver.
+type DNSNameResolverSpec struct {
+	// name is the DNS name for which the DNS name resolution information will be stored.
+	// For a regular DNS name, only the DNS name resolution information of the regular DNS
+	// name will be stored. For a wildcard DNS name, the DNS name resolution information
+	// of all the DNS names, that matches the wildcard DNS name, will be stored.
+	// For a wildcard DNS name, the '*' will match only one label. Additionally, only a single
+	// '*' can be used at the beginning of the wildcard DNS name. For example, '*.example.com.'
+	// will match 'sub1.example.com.' but won't match 'sub2.sub1.example.com.'
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.)+[A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.$
+	// +kubebuilder:validation:MaxLength=254
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec.name is immutable"
+	Name string `json:"name"`
+}
+
+// DNSNameResolverStatus defines the observed status of DNSNameResolver.
+type DNSNameResolverStatus struct {
+	// resolvedNames contains a list of matching DNS names and their corresponding IP addresses
+	// along with their TTL and last DNS lookup times.
+	// +listType=map
+	// +listMapKey=dnsName
+	// +patchMergeKey=dnsName
+	// +patchStrategy=merge
+	// +optional
+	ResolvedNames []DNSNameResolverResolvedName `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
+}
+
+// DNSNameResolverResolvedName describes the details of a resolved DNS name.
+type DNSNameResolverResolvedName struct {
+	// conditions provide information about the state of the DNS name.
+	// Known .status.conditions.type is: "Degraded"
```

Comment:
Can you list here, under what circumstances is the status set to Degraded=true?

Reply:
It was previously suggested to remove the details here: https://github.com/openshift/api/pull/1524#discussion_r1276251400

Reply:
The status will be set to Degraded=true whenever a DNS name resolution fails. It will be set to Degarded=false when the DNS name resolution succeeds.

Reply:
It is easier to get this information into the documentation if it exists in the godoc.  I think all you have to say is "Degraded is true when there is a DNS name resolution failure, and false when there is no failure."

Reply:
Added the details in the godoc.